tfe/resource_tfe_workspace_test.go

```diff
@@ -362,6 +366,42 @@ func TestAccTFEWorkspace_updateTriggerPrefixes(t *testing.T) {
 	})
 }
 
+func TestAccTFEWorkspace_changeTags(t *testing.T) {
+	workspace := &tfe.Workspace{}
+	rInt := rand.New(rand.NewSource(time.Now().UnixNano())).Int()
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckTFEWorkspaceDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccTFEWorkspace_basic(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckTFEWorkspaceExists(
+						"tfe_workspace.foobar", workspace),
+					resource.TestCheckResourceAttr(
+						"tfe_workspace.foobar", "tag_names.0", "fav"),
+					resource.TestCheckResourceAttr(
+						"tfe_workspace.foobar", "tag_names.1", "test"),
+				),
+			},
+
+			{
+				Config: testAccTFEWorkspace_basicChangeTags(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckTFEWorkspaceExists(
+						"tfe_workspace.foobar", workspace),
+					resource.TestCheckResourceAttr(
+						"tfe_workspace.foobar", "tag_names.0", "prod"),
+					resource.TestCheckResourceAttr(
+						"tfe_workspace.foobar", "tag_names.1", "unfav"),
+				),
+			},
+		},
+	})
+}
```

Comment:
Could we add some more complicated tests to handle the difference check where we add/remove tags?

1. A test where we create a workspace with 0 tags and then update that same workspace to add a few tags
2. A test where we create a workspace with  2 tags and then do an update on a different field (say name) and ensure all the tags are present
3. A test where we create a workspace with  2 tags and then remove 1 tag
4. A test where we create a workspace with  2 tags and then change the name of one tag and ensure there is a removal then a new one. 

Sometimes in [other tests](https://github.com/hashicorp/terraform-provider-tfe/blob/main/tfe/data_source_outputs_test.go#L113) we've used the tfeClient to perform some ancillary tasks. Here, we could write some extra test verification functions to ensure that the tags are not present on a workspace.

Reply:
i found a sequence of tests that should cover each of those. they don't start from scratch each time though--is there something I miss by not doing that?

Reply:
That is fine 👍